### PR TITLE
Adds opt-in rule discouraged_optional_collection

### PR DESCRIPTION
**Pull request by ornithocoder**
Implements #1885.

I tried to use regular expressions to match collections, but soon realized I would need recursion support - `?R` - to find and match nested collections, something Foundation doesn't offer.

So I decided to re-implement the rule as an `ASTRule`, looking for balanced optional collections in functions and variables declarations.

Example: `[String: [Int]?]`

```
[  S  t  r  i  n  g  :     [  I  n  t  ]  ?  ]
0  1  2  3  4  5  6  7  8  9  10 11 12 13 14 15
                           ^              ^
// [9, 14]
```

Example: `[String: [Int]?]?`

```
[  S  t  r  i  n  g  :     [  I  n  t  ]  ?  ]  ?
0  1  2  3  4  5  6  7  8  9  10 11 12 13 14 15 16
^                          ^              ^     ^

// [0, 16], [9, 14]
```

Example: `var x = Set<Int>?`

```
v  a  r     x     =     S  e  t  <  I  n  t  >  ?
0  1  2  3  4  5  6  7  8  9  10 11 12 13 14 15 16
                        ^                       ^
// [8, 16]
```

I've added similar examples to the private `String` extension used by the rule to illustrate the code.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,20 @@
+## Master
+
+#### Breaking
+
+* None.
+
+#### Enhancements
+
+* Adds `discouraged_optional_collection` opt-in rule to encourages the return
```
Comment (jpsim) [inline CHANGELOG.md]: s/encourages the return/encourage the use/

Comment (ornithocoder) [reply CHANGELOG.md]: Thanks, @jpsim! 

I'll have to refactor/rewrite this rule. There are some cases that are really hard to cover using regular expression. For instance:

```swift
fun foo(param: [String: ↓[String: String]?])
```

vs.

```swift
fun foo(param: ↓[String: [String: String]]?)
```

But thanks, I'll update the changelog entry.

**Comment by ornithocoder** [comment]
Hi @marcelofabri / @jpsim. Did you have a chance to look into this one? I added comments to the code to illustrate it better. For "developers of the future" that might need to interact with the code. Thank!

**Comment by ornithocoder** [comment]
I created a [gist with the String extension](https://gist.github.com/ornithocoder/34cef3db17edb2cf88c1cae7b160f305) from this rule. It can be copied and pasted on a Playground.

**Comment by ornithocoder** [comment]
Hi @marcelofabri, I am sorry to bother you with this but can you take a look on this PR? I've been using it on my project at work for weeks now and it seems solid. I'd love to see it merged into SwiftLint's master. Thanks!

**Comment by marcelofabri** [comment]
Thanks @ornithocoder 💯

**Pull request merged**